pyteal/ast/abi/type.py

```diff
@@ -182,3 +182,40 @@ def use(self, action: Callable[[T], Expr]) -> Expr:
 
 
 ComputedValue.__module__ = "pyteal"
+
+
+class ReturnedValue(ComputedValue):
+    def __init__(self, type_spec: TypeSpec, computation_expr: Expr):
+        from pyteal.ast.subroutine import SubroutineCall
+
+        self.type_spec = type_spec
+        if not isinstance(computation_expr, SubroutineCall):
+            raise TealInputError(
+                f"Expecting computation_expr to be SubroutineCall but get {type(computation_expr)}"
+            )
+        self.computation = computation_expr
+
+    def produced_type_spec(self) -> TypeSpec:
+        return self.type_spec
+
+    def store_into(self, output: BaseType) -> Expr:
+        if output.type_spec() != self.produced_type_spec():
+            raise TealInputError(
+                f"expected type_spec {self.produced_type_spec()} but get {output.type_spec()}"
+            )
+
+        declaration = self.computation.subroutine.get_declaration()
+
+        if declaration.deferred_expr is None:
+            raise TealInputError(
+                "ABI return subroutine must have deferred_expr to be not-None."
+            )
+        if declaration.deferred_expr.type_of() != output.type_spec().storage_type():
+            raise TealInputError(
```

Comment:
I think the check is needed.

Why we need to trouble `stored_value.slot.store` not `stored_value.store`?
Since `self.computation` is evaluated to `TealType.none`, but latter option does the type check to align the type of `self.computation` to scratchvar's type.

The "runtime behavior of" `self.computation` type is ABI's scratch var's type, which is equal to `deferred_expr`'s type, and this is the missing type check in `stored_value.slot.store`.

Reply:
Great point, I misunderstood this. Resolving